mars/tensor/expressions/arithmetic/clip.py

```diff
@@ -51,80 +51,26 @@ def a_max(self):
     def out(self):
         return getattr(self, '_out', None)
 
-    @contextlib.contextmanager
-    def _handle_params(self, inputs):
-        inps = inputs[:1]
-        inputs_iter = iter(inputs[1:])
+    def _set_inputs(self, inputs):
+        super(TensorClip, self)._set_inputs(inputs)
+        inputs_iter = iter(self._inputs)
 
         if getattr(self, '_a', None) is None:
             # create clip op from beginning
-
-            # a_min
-            has_a_min = False
-            a_min = next(inputs_iter)
-            if a_min is not None and isinstance(a_min, TENSOR_TYPE + CHUNK_TYPE):
-                has_a_min = True
-                inps.append(a_min)
-            else:
-                setattr(self, '_a_min', a_min)
-
-            # a_max
-            has_a_max = False
-            a_max = next(inputs_iter)
-            if a_max is not None and isinstance(a_max, TENSOR_TYPE + CHUNK_TYPE):
-                has_a_max = True
-                inps.append(a_max)
-            else:
-                setattr(self, '_a_max', a_max)
-
-            # out
-            has_out = False
-            out = next(inputs_iter)
-            if out is not None:
-                has_out = True
-                inps.append(out)
+            self._a = next(inputs_iter)
+            self._a_min = next(inputs_iter) if isinstance(inputs[1], (Base, Entity)) else inputs[1]
+            self._a_max = next(inputs_iter) if isinstance(inputs[2], (Base, Entity)) else inputs[2]
+            if inputs[3] is not None:
+                self._out = next(inputs_iter)
```

Comment:
When either inputs[1] or inputs[2] is neither Base nor Entity, self._out will be inputs[1]. Is this behavior expected?

Reply:
`self._inputs` are all `Base` or `Entity`, if inputs[1] or inputs[2] is neither Base nor Entity, and `self._out` is not `None`, then `self._inputs` has only two elements.

> self._out will be inputs[1] while we just checked inputs[3]

`self._out` will be `self._inputs[1]`

Reply:
So why we need to check inputs[3] while we actually deal with inputs[1]?

Reply:
it's self._inputs[1], not the inputs[1], they are different after calling the `super(TensorClip, self)._set_inputs(inputs)`.

Reply:
Therefore it is better to reference inputs in a unified manner, for instance, use inputs[idx] instead of mixing self._inputs, inputs and inputs_iter.